File: 01/01.py
from functools import partial
from typing import Sequence


def part1(data: Sequence[int], *, target: int = 2020) -> int:
    for a in data:
        for b in data:
            if a + b == target:
                return a * b


def part2(data: Sequence[int], *, target: int = 2020) -> int:
    for a in data:
        for b in data:
            if (part1 := a + b) >= target:
                continue
            for c in data:
                if part1 + c == target:
                    return a * b * c


example1 = partial(part1, data=[1721, 979, 366, 299, 675, 1456])
example2 = partial(part2, data=[1721, 979, 366, 299, 675, 1456])


if __name__ == '__main__':
    example1() == 876459
    example2() == 116168640
    
    with open('01/input.txt', 'r') as in_file:
        data = list(map(int, in_file))
    
    print(part1(data))
    print(part2(data))

File: 01/01.c
# include <stdio.h>
# include <stdlib.h>
# include "../util.c"

# define ARR_SIZE 200

long part1(struct array_with_length* data, int target) {
    long* vals = data->data;
    for (int i = 0; i < data->len; i++) {
        for (int j = 0; j < data->len; j++) {
            if (vals[i] + vals[j] == target) {
                return vals[i] * vals[j];
            }
        }
    }
    return -1;
}

long part2(struct array_with_length* data, int target) {
    long* vals = data->data;
    for (int i = 0; i < data->len; i++) {
        for (int j = 0; j < data->len; j++) {
            for (int k = 0; k < data->len; k++) {
                if (vals[i] + vals[j] + vals[k] == target) {
                    return vals[i] * vals[j] * vals[k];
                }
            }
        }
    }
    return -1;
}

int main() {
    struct array_with_length* data = read_input("01/input.txt", ARR_SIZE);
    if (data == NULL) {
        return EXIT_FAILURE;
    }

    const long p1 = part1(data, 2020);
    printf("part1: %ld\n", p1);

    const long p2 = part2(data, 2020);
    printf("part2: %ld\n", p2);

    free(data->data);
    free(data);
    exit(EXIT_SUCCESS);
}
